File: test/compare/compare_nifti.py
#!/usr/bin/env python3
"""Compare two NIfTI files and report detailed statistics on differences.

Usage:
    python compare_nifti.py file1.nii file2.nii [--tolerance 1e-6] [--json] [--verbose]
    python compare_nifti.py --dir rust_output/ julia_output/ [--tolerance 1e-6] [--json]

Exit codes:
    0 — Files match within tolerance
    1 — Files differ beyond tolerance
    2 — Error (file not found, incompatible shapes, etc.)
"""

import argparse
import json
import os
import struct
import sys
import gzip
from pathlib import Path

# ── NIfTI-1 minimal reader (no external dependencies) ──────────────────────

NIFTI1_HEADER_SIZE = 348


def _open_nifti(path):
    """Return a file-like object for .nii or .nii.gz."""
    path = str(path)
    if path.endswith(".gz"):
        return gzip.open(path, "rb")
    return open(path, "rb")


def _read_header(fp):
    """Parse NIfTI-1 header fields needed for comparison."""
    hdr_bytes = fp.read(NIFTI1_HEADER_SIZE)
    if len(hdr_bytes) < NIFTI1_HEADER_SIZE:
        raise ValueError("File too small for NIfTI-1 header")

    sizeof_hdr = struct.unpack_from("<i", hdr_bytes, 0)[0]
    if sizeof_hdr != 348:
        raise ValueError(f"Invalid NIfTI-1 header size: {sizeof_hdr}")

    dim = struct.unpack_from("<8h", hdr_bytes, 40)
    ndim = dim[0]
    shape = tuple(dim[1 : ndim + 1])

    datatype = struct.unpack_from("<h", hdr_bytes, 70)[0]
    bitpix = struct.unpack_from("<h", hdr_bytes, 72)[0]

    pixdim = struct.unpack_from("<8f", hdr_bytes, 76)
    voxel_size = tuple(pixdim[1 : ndim + 1])

    vox_offset = struct.unpack_from("<f", hdr_bytes, 108)[0]
    scl_slope = struct.unpack_from("<f", hdr_bytes, 112)[0]
    scl_inter = struct.unpack_from("<f", hdr_bytes, 116)[0]

    return {
        "shape": shape,
        "datatype": datatype,
        "bitpix": bitpix,
        "voxel_size": voxel_size,
        "vox_offset": int(vox_offset),
        "scl_slope": scl_slope,
        "scl_inter": scl_inter,
    }


# Mapping NIfTI datatype codes → struct format characters
_DTYPE_MAP = {
    2: ("B", 1),  # uint8
    4: ("h", 2),  # int16
    8: ("i", 4),  # int32
    16: ("f", 4),  # float32
    32: ("d", 8),  # float64
    64: ("d", 8),  # float64 (complex64 pair — rare)
    256: ("b", 1),  # int8
    512: ("H", 2),  # uint16
    768: ("I", 4),  # uint32
}


def _read_data(fp, header):
    """Read voxel data as a flat list of floats."""
    fp.seek(header["vox_offset"])
    dtype_code = header["datatype"]
    if dtype_code not in _DTYPE_MAP:
        raise ValueError(f"Unsupported NIfTI datatype code: {dtype_code}")

    fmt_char, byte_size = _DTYPE_MAP[dtype_code]
    nvoxels = 1
    for s in header["shape"]:
        nvoxels *= s

    raw = fp.read(nvoxels * byte_size)
    if len(raw) < nvoxels * byte_size:
        raise ValueError(
            f"Data truncated: expected {nvoxels * byte_size} bytes, got {len(raw)}"
        )

    data = list(struct.unpack(f"<{nvoxels}{fmt_char}", raw))

    slope = header["scl_slope"]
    inter = header["scl_inter"]
    if slope != 0 and (slope != 1.0 or inter != 0.0):
        data = [v * slope + inter for v in data]

    return data


def read_nifti(path):
    """Read a NIfTI file and return (header_dict, flat_data_list)."""
    with _open_nifti(path) as fp:
        header = _read_header(fp)
        data = _read_data(fp, header)
    return header, data


# ── Comparison logic ────────────────────────────────────────────────────────


def compare_data(data1, data2, tolerance):
    """Compute comparison statistics between two flat data arrays."""
    n = len(data1)
    if n == 0:
        return {"error": "Empty data arrays"}

    abs_diffs = [abs(a - b) for a, b in zip(data1, data2)]
    max_abs_diff = max(abs_diffs)
    mean_abs_diff = sum(abs_diffs) / n
    rmse = (sum(d * d for d in abs_diffs) / n) ** 0.5

    exact_match = sum(1 for d in abs_diffs if d == 0.0)
    within_tol = sum(1 for d in abs_diffs if d <= tolerance)

    # Data range for normalization
    all_vals = data1 + data2
    vmin = min(all_vals)
    vmax = max(all_vals)
    data_range = vmax - vmin
    nrmse = rmse / data_range if data_range > 0 else 0.0

    # Pearson correlation
    mean1 = sum(data1) / n
    mean2 = sum(data2) / n
    cov = sum((a - mean1) * (b - mean2) for a, b in zip(data1, data2)) / n
    std1 = (sum((a - mean1) ** 2 for a in data1) / n) ** 0.5
    std2 = (sum((b - mean2) ** 2 for b in data2) / n) ** 0.5
    if std1 > 0 and std2 > 0:
        correlation = cov / (std1 * std2)
    else:
        correlation = 1.0 if std1 == 0 and std2 == 0 else 0.0

    # Distribution of differences
    thresholds = [1e-10, 1e-7, 1e-5, 1e-3, 1e-1, 1.0]
    diff_distribution = {}
    for t in thresholds:
        count = sum(1 for d in abs_diffs if d <= t)
        diff_distribution[f"<={t:.0e}"] = count

    return {
        "n_voxels": n,
        "max_abs_diff": max_abs_diff,
        "mean_abs_diff": mean_abs_diff,
        "rmse": rmse,
        "nrmse": nrmse,
        "correlation": correlation,
        "exact_match_count": exact_match,
        "exact_match_pct": 100.0 * exact_match / n,
        "within_tolerance_count": within_tol,
        "within_tolerance_pct": 100.0 * within_tol / n,
        "data_range": data_range,
        "diff_distribution": diff_distribution,
        "pass": max_abs_diff <= tolerance,
    }


def compare_files(file1, file2, tolerance, verbose=False):
    """Compare two NIfTI files. Returns a result dict."""
    result = {
        "file1": str(file1),
        "file2": str(file2),
        "tolerance": tolerance,
    }

    try:
        hdr1, data1 = read_nifti(file1)
        hdr2, data2 = read_nifti(file2)
    except Exception as e:
        result["error"] = str(e)
        result["pass"] = False
        return result

    # Header comparison
    result["shape1"] = hdr1["shape"]
    result["shape2"] = hdr2["shape"]
    result["datatype1"] = hdr1["datatype"]
    result["datatype2"] = hdr2["datatype"]
    result["voxel_size1"] = hdr1["voxel_size"]
    result["voxel_size2"] = hdr2["voxel_size"]

    if hdr1["shape"] != hdr2["shape"]:
        result["error"] = (
            f"Shape mismatch: {hdr1['shape']} vs {hdr2['shape']}"
        )
        result["pass"] = False
        return result

    # Data comparison
    stats = compare_data(data1, data2, tolerance)
    result.update(stats)

    return result


def format_result(result, verbose=False):
    """Format comparison result as a human-readable string."""
    lines = []
    f1 = os.path.basename(result.get("file1", "?"))
    f2 = os.path.basename(result.get("file2", "?"))
    lines.append(f"  Comparing: {f1}  vs  {f2}")

    if "error" in result:
        lines.append(f"  ERROR: {result['error']}")
        return "\n".join(lines)

    passed = result.get("pass", False)
    status = "PASS ✓" if passed else "FAIL ✗"
    lines.append(f"  Status: {status}")
    lines.append(f"  Shape: {result['shape1']}")

    if result.get("datatype1") != result.get("datatype2"):
        lines.append(
            f"  Datatype: {result['datatype1']} vs {result['datatype2']} (DIFFERENT)"
        )

    lines.append(f"  Max absolute diff:  {result['max_abs_diff']:.6e}")
    lines.append(f"  Mean absolute diff: {result['mean_abs_diff']:.6e}")
    lines.append(f"  RMSE:               {result['rmse']:.6e}")
    lines.append(f"  Normalized RMSE:    {result['nrmse']:.6e}")
    lines.append(f"  Correlation:        {result['correlation']:.10f}")
    lines.append(
        f"  Exact match:        {result['exact_match_count']}/{result['n_voxels']} "
        f"({result['exact_match_pct']:.2f}%)"
    )
    lines.append(
        f"  Within tolerance:   {result['within_tolerance_count']}/{result['n_voxels']} "
        f"({result['within_tolerance_pct']:.2f}%)"
    )

    if verbose and "diff_distribution" in result:
        lines.append("  Difference distribution:")
        for threshold, count in result["diff_distribution"].items():
            pct = 100.0 * count / result["n_voxels"]
            lines.append(f"    {threshold}: {count} ({pct:.2f}%)")

    return "\n".join(lines)


# ── Directory comparison ────────────────────────────────────────────────────


def find_nifti_files(directory):
    """Find all NIfTI files in a directory (non-recursive)."""
    nifti_files = {}
    for f in os.listdir(directory):
        if f.endswith(".nii") or f.endswith(".nii.gz"):
            nifti_files[f] = os.path.join(directory, f)
    return nifti_files


def compare_directories(dir1, dir2, tolerance, verbose=False):
    """Compare all matching NIfTI files between two directories."""
    files1 = find_nifti_files(dir1)
    files2 = find_nifti_files(dir2)

    all_names = sorted(set(files1.keys()) | set(files2.keys()))
    results = []

    for name in all_names:
        if name.startswith("settings_"):
            continue  # skip settings files

        if name not in files1:
            results.append(
                {
                    "file1": f"MISSING ({name})",
                    "file2": files2[name],
                    "error": f"File only in {dir2}",
                    "pass": False,
                }
            )
        elif name not in files2:
            results.append(
                {
                    "file1": files1[name],
                    "file2": f"MISSING ({name})",
                    "error": f"File only in {dir1}",
                    "pass": False,
                }
            )
        else:
            result = compare_files(files1[name], files2[name], tolerance, verbose)
            results.append(result)

    return results


# ── CLI ─────────────────────────────────────────────────────────────────────


def main():
    parser = argparse.ArgumentParser(
        description="Compare NIfTI files between Rust and Julia implementations",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )

    parser.add_argument(
        "paths",
        nargs="*",
        help="Two NIfTI files or two directories to compare",
    )
    parser.add_argument(
        "--dir",
        action="store_true",
        help="Compare all matching NIfTI files in two directories",
    )
    parser.add_argument(
        "--tolerance",
        type=float,
        default=1e-6,
        help="Maximum allowed absolute difference (default: 1e-6)",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output results as JSON",
    )
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Show detailed difference distribution",
    )

    args = parser.parse_args()

    if len(args.paths) != 2:
        parser.error("Exactly two paths required (files or directories)")

    path1, path2 = args.paths

    if args.dir or (os.path.isdir(path1) and os.path.isdir(path2)):
        results = compare_directories(path1, path2, args.tolerance, args.verbose)
    else:
        results = [compare_files(path1, path2, args.tolerance, args.verbose)]

    # Output
    if args.json:
        # Convert tuples to lists for JSON serialization
        def make_serializable(obj):
            if isinstance(obj, dict):
                return {k: make_serializable(v) for k, v in obj.items()}
            if isinstance(obj, (list, tuple)):
                return [make_serializable(i) for i in obj]
            return obj

        print(json.dumps(make_serializable(results), indent=2))
    else:
        all_pass = True
        for r in results:
            print(format_result(r, args.verbose))
            print()
            if not r.get("pass", False):
                all_pass = False

        n_pass = sum(1 for r in results if r.get("pass", False))
        n_total = len(results)
        print(f"{'=' * 60}")
        print(f"Summary: {n_pass}/{n_total} comparisons passed (tolerance={args.tolerance:.0e})")
        if all_pass:
            print("Overall: PASS ✓")
        else:
            print("Overall: FAIL ✗")

    # Exit code
    all_pass = all(r.get("pass", False) for r in results)
    sys.exit(0 if all_pass else 1)


if __name__ == "__main__":
    main()

File: test/compare/julia/compare_nifti.jl
#!/usr/bin/env julia
#=
Compare two NIfTI files and report detailed statistics on differences.

Usage:
    julia --project=test/compare/julia compare_nifti.jl file1.nii file2.nii [--tolerance 1e-6] [--json] [--verbose]
    julia --project=test/compare/julia compare_nifti.jl --dir rust_output/ julia_output/ [--tolerance 1e-6] [--json]

Exit codes:
    0 — Files match within tolerance
    1 — Files differ beyond tolerance
    2 — Error (file not found, incompatible shapes, etc.)
=#

using NIfTI
using Statistics
using Printf
using JSON3

# ── Comparison logic ────────────────────────────────────────────────────────

function compare_data(data1::AbstractArray, data2::AbstractArray, tolerance::Float64)
    v1 = Float64.(vec(data1))
    v2 = Float64.(vec(data2))
    n = length(v1)
    n == 0 && return Dict("error" => "Empty data arrays")

    # Ignore voxels where either side is NaN (Julia marks out-of-mask as
    # NaN via gaussiansmooth3d's mask handling; Rust leaves them at the
    # input value). Statistics are over the in-mask intersection.
    valid = .!isnan.(v1) .& .!isnan.(v2)
    n_nan_only_in_1 = count(isnan.(v1) .& .!isnan.(v2))
    n_nan_only_in_2 = count(.!isnan.(v1) .& isnan.(v2))
    n_valid = count(valid)

    if n_valid == 0
        return Dict("error" => "No overlapping non-NaN voxels", "pass" => false)
    end

    vv1 = v1[valid]
    vv2 = v2[valid]

    abs_diffs = abs.(vv1 .- vv2)
    max_abs_diff = maximum(abs_diffs)
    mean_abs_diff = mean(abs_diffs)
    rmse = sqrt(mean(abs_diffs .^ 2))

    exact_match = count(==(0.0), abs_diffs)
    within_tol = count(<=(tolerance), abs_diffs)

    # Data range for normalization
    all_vals = vcat(vv1, vv2)
    data_range = maximum(all_vals) - minimum(all_vals)
    nrmse = data_range > 0 ? rmse / data_range : 0.0

    # Pearson correlation
    corr = if std(vv1) > 0 && std(vv2) > 0
        cor(vv1, vv2)
    else
        std(vv1) == 0 && std(vv2) == 0 ? 1.0 : 0.0
    end

    # Distribution of differences
    thresholds = [1e-10, 1e-7, 1e-5, 1e-3, 1e-1, 1.0]
    diff_distribution = Dict{String,Int}()
    for t in thresholds
        key = string("<=", @sprintf("%.0e", t))
        diff_distribution[key] = count(<=(t), abs_diffs)
    end

    # In-mask subset: also drop voxels where either side is exactly 0.0.
    # Rust binaries leave out-of-mask voxels at 0 (their original input
    # value); Julia's NaN-marking only catches the smoothing-affected
    # subset. Adding "both != 0" removes the (Rust=0, Julia=random-noise)
    # pairs that otherwise drag the correlation down. Safe for phase,
    # mask, B0 outputs; for SWI/MIP a true 0 voxel is rare in practice.
    inmask = valid .& (v1 .!= 0.0) .& (v2 .!= 0.0)
    n_inmask = count(inmask)
    inmask_stats = if n_inmask > 0
        v1im = v1[inmask]
        v2im = v2[inmask]
        ad = abs.(v1im .- v2im)
        corr_im = if std(v1im) > 0 && std(v2im) > 0; cor(v1im, v2im); else 1.0; end
        Dict(
            "inmask_n" => n_inmask,
            "inmask_max_abs_diff" => maximum(ad),
            "inmask_mean_abs_diff" => mean(ad),
            "inmask_correlation" => corr_im,
            "inmask_within_tol_pct" => 100.0 * count(<=(tolerance), ad) / n_inmask,
        )
    else
        Dict("inmask_n" => 0)
    end

    base = Dict(
        "n_voxels" => n,
        "n_valid" => n_valid,
        "n_nan_only_in_1" => n_nan_only_in_1,
        "n_nan_only_in_2" => n_nan_only_in_2,
        "max_abs_diff" => max_abs_diff,
        "mean_abs_diff" => mean_abs_diff,
        "rmse" => rmse,
        "nrmse" => nrmse,
        "correlation" => corr,
        "exact_match_count" => exact_match,
        "exact_match_pct" => 100.0 * exact_match / n_valid,
        "within_tolerance_count" => within_tol,
        "within_tolerance_pct" => 100.0 * within_tol / n_valid,
        "data_range" => data_range,
        "diff_distribution" => diff_distribution,
        "pass" => max_abs_diff <= tolerance,
    )
    return merge(base, inmask_stats)
end

function compare_files(file1::String, file2::String, tolerance::Float64)
    result = Dict{String,Any}(
        "file1" => file1,
        "file2" => file2,
        "tolerance" => tolerance,
    )

    local nii1, nii2
    try
        nii1 = niread(file1)
        nii2 = niread(file2)
    catch e
        result["error"] = string(e)
        result["pass"] = false
        return result
    end

    shape1 = size(nii1)
    shape2 = size(nii2)
    result["shape1"] = collect(shape1)
    result["shape2"] = collect(shape2)
    result["datatype1"] = Int(nii1.header.datatype)
    result["datatype2"] = Int(nii2.header.datatype)
    result["voxel_size1"] = collect(Float64.(nii1.header.pixdim[2:ndims(nii1)+1]))
    result["voxel_size2"] = collect(Float64.(nii2.header.pixdim[2:ndims(nii2)+1]))

    if shape1 != shape2
        result["error"] = "Shape mismatch: $shape1 vs $shape2"
        result["pass"] = false
        return result
    end

    stats = compare_data(nii1.raw, nii2.raw, tolerance)
    merge!(result, stats)
    return result
end

# ── Formatting ──────────────────────────────────────────────────────────────

function format_result(result::Dict; verbose::Bool=false)
    lines = String[]
    f1 = basename(get(result, "file1", "?"))
    f2 = basename(get(result, "file2", "?"))
    push!(lines, "  Comparing: $f1  vs  $f2")

    if haskey(result, "error")
        push!(lines, "  ERROR: $(result["error"])")
        return join(lines, "\n")
    end

    passed = get(result, "pass", false)
    status = passed ? "PASS ✓" : "FAIL ✗"
    push!(lines, "  Status: $status")
    push!(lines, "  Shape: $(result["shape1"])")

    if get(result, "datatype1", 0) != get(result, "datatype2", 0)
        push!(lines, "  Datatype: $(result["datatype1"]) vs $(result["datatype2"]) (DIFFERENT)")
    end

    push!(lines, @sprintf("  Max absolute diff:  %.6e", result["max_abs_diff"]))
    push!(lines, @sprintf("  Mean absolute diff: %.6e", result["mean_abs_diff"]))
    push!(lines, @sprintf("  RMSE:               %.6e", result["rmse"]))
    push!(lines, @sprintf("  Normalized RMSE:    %.6e", result["nrmse"]))
    push!(lines, @sprintf("  Correlation:        %.10f", result["correlation"]))
    push!(lines, @sprintf("  Exact match:        %d/%d (%.2f%%)",
        result["exact_match_count"], result["n_voxels"], result["exact_match_pct"]))
    n_valid = get(result, "n_valid", result["n_voxels"])
    push!(lines, @sprintf("  Within tolerance:   %d/%d (%.2f%%)",
        result["within_tolerance_count"], n_valid, result["within_tolerance_pct"]))
    if get(result, "n_nan_only_in_1", 0) > 0 || get(result, "n_nan_only_in_2", 0) > 0
        push!(lines, @sprintf("  NaN-only voxels:    %d in left, %d in right (excluded from stats)",
            result["n_nan_only_in_1"], result["n_nan_only_in_2"]))
    end
    if get(result, "inmask_n", 0) > 0 && result["inmask_n"] != n_valid
        push!(lines, @sprintf("  In-mask subset (both ≠ 0): %d voxels", result["inmask_n"]))
        push!(lines, @sprintf("    max abs diff:  %.6e", result["inmask_max_abs_diff"]))
        push!(lines, @sprintf("    correlation:   %.6f", result["inmask_correlation"]))
        push!(lines, @sprintf("    within tol:    %.2f%%", result["inmask_within_tol_pct"]))
    end

    if verbose && haskey(result, "diff_distribution")
        push!(lines, "  Difference distribution:")
        for (threshold, count) in sort(collect(result["diff_distribution"]))
            pct = 100.0 * count / result["n_voxels"]
            push!(lines, @sprintf("    %s: %d (%.2f%%)", threshold, count, pct))
        end
    end

    return join(lines, "\n")
end

# ── Directory comparison ────────────────────────────────────────────────────

function find_nifti_files(directory::String)
    files = Dict{String,String}()
    for f in readdir(directory)
        if endswith(f, ".nii") || endswith(f, ".nii.gz")
            files[f] = joinpath(directory, f)
        end
    end
    return files
end

function compare_directories(dir1::String, dir2::String, tolerance::Float64)
    files1 = find_nifti_files(dir1)
    files2 = find_nifti_files(dir2)

    all_names = sort(collect(union(keys(files1), keys(files2))))
    results = Dict{String,Any}[]

    for name in all_names
        startswith(name, "settings_") && continue

        if !haskey(files1, name)
            push!(results, Dict{String,Any}(
                "file1" => "MISSING ($name)",
                "file2" => files2[name],
                "error" => "File only in $dir2",
                "pass" => false,
            ))
        elseif !haskey(files2, name)
            push!(results, Dict{String,Any}(
                "file1" => files1[name],
                "file2" => "MISSING ($name)",
                "error" => "File only in $dir1",
                "pass" => false,
            ))
        else
            push!(results, compare_files(files1[name], files2[name], tolerance))
        end
    end

    return results
end

# ── JSON output ─────────────────────────────────────────────────────────────

function to_json(results)
    return JSON3.pretty(results)
end

# ── CLI ─────────────────────────────────────────────────────────────────────

function parse_cli_args(args)
    tolerance = 1e-6
    json_output = false
    verbose = false
    dir_mode = false
    paths = String[]

    i = 1
    while i <= length(args)
        arg = args[i]
        if arg == "--tolerance" && i < length(args)
            tolerance = parse(Float64, args[i+1])
            i += 2
        elseif arg == "--json"
            json_output = true
            i += 1
        elseif arg == "--verbose" || arg == "-v"
            verbose = true
            i += 1
        elseif arg == "--dir"
            dir_mode = true
            i += 1
        elseif arg == "--help" || arg == "-h"
            println("""
Compare NIfTI files between Rust and Julia implementations.

Usage:
    julia compare_nifti.jl file1.nii file2.nii [--tolerance 1e-6] [--json] [--verbose]
    julia compare_nifti.jl --dir dir1/ dir2/ [--tolerance 1e-6] [--json] [--verbose]

Options:
    --tolerance TOL  Maximum allowed absolute difference (default: 1e-6)
    --json           Output results as JSON
    --verbose, -v    Show detailed difference distribution
    --dir            Compare all matching NIfTI files in two directories
    --help, -h       Show this help
""")
            exit(0)
        else
            push!(paths, arg)
            i += 1
        end
    end

    return (; tolerance, json_output, verbose, dir_mode, paths)
end

function main()
    opts = parse_cli_args(ARGS)

    if length(opts.paths) != 2
        println(stderr, "error: exactly two paths required (files or directories)")
        exit(2)
    end

    path1, path2 = opts.paths
    dir_mode = opts.dir_mode || (isdir(path1) && isdir(path2))

    results = if dir_mode
        compare_directories(path1, path2, opts.tolerance)
    else
        [compare_files(path1, path2, opts.tolerance)]
    end

    if opts.json_output
        println(to_json(results))
    else
        all_pass = true
        for r in results
            println(format_result(r; verbose=opts.verbose))
            println()
            if !get(r, "pass", false)
                all_pass = false
            end
        end

        n_pass = count(r -> get(r, "pass", false), results)
        n_total = length(results)
        println("=" ^ 60)
        @printf("Summary: %d/%d comparisons passed (tolerance=%.0e)\n", n_pass, n_total, opts.tolerance)
        if all_pass
            println("Overall: PASS ✓")
        else
            println("Overall: FAIL ✗")
        end
    end

    has_errors = any(r -> haskey(r, "error"), results)
    all_pass = all(r -> get(r, "pass", false), results)
    exit(has_errors ? 2 : (all_pass ? 0 : 1))
end

main()
